packages/protocol/contracts/layer1/based2/libs/LibValidate.sol

```diff
@@ -0,0 +1,381 @@
+    /// @param _signalSlots Array of signal slot identifiers to validate
+    function _validateSignals(
+        I.Config memory _conf,
+        LibState.ReadWrite memory _rw,
+        I.Block[] memory _blocks,
+        bytes32[] memory _signalSlots
+    )
+        internal
+        view
+    {
+        unchecked {
+            uint256 k;
+
+            for (uint256 i; i < _blocks.length; ++i) {
+                if (_blocks[i].numSignals == 0) continue;
+
+                require(_blocks[i].numSignals <= _conf.maxSignalsToReceive, TooManySignals());
+
+                for (uint256 j; j < _blocks[i].numSignals; ++j) {
+                    require(_rw.isSignalSent(_conf, _signalSlots[k++]), SignalNotSent());
+                }
+            }
+        }
+    }
+
+    /// @notice Validates anchor blocks used for L1-L2 synchronization
+    /// @dev Ensures anchor blocks are properly ordered, within height limits, and have valid hashes
+    /// @param _conf Protocol configuration
+    /// @param _rw Read/write access functions
+    /// @param _batch The batch being validated
+    /// @param _blocks Array of blocks in the batch
+    /// @param _parentLastAnchorBlockId Last anchor block ID from parent batch
+    /// @return anchorBlockHashes_ Array of validated anchor block hashes
+    /// @return lastAnchorBlockId_ ID of the last anchor block in this batch
+    function _validateAnchors(
+        I.Config memory _conf,
+        LibState.ReadWrite memory _rw,
+        I.Batch calldata _batch,
+        I.Block[] memory _blocks,
+        uint48 _parentLastAnchorBlockId
+    )
+        internal
+        view
+        returns (bytes32[] memory anchorBlockHashes_, uint48 lastAnchorBlockId_)
+    {
+        anchorBlockHashes_ = new bytes32[](_blocks.length);
+        lastAnchorBlockId_ = _parentLastAnchorBlockId;
+
+        uint256 k;
+        bool anchorFound;
+
+        for (uint256 i; i < _blocks.length; ++i) {
+            if (!_blocks[i].hasAnchor) continue;
+
+            require(k < _batch.anchorBlockIds.length, NotEnoughAnchorIds());
+
+            uint48 anchorBlockId = _batch.anchorBlockIds[k];
+            require(anchorBlockId != 0, AnchorIdZero());
+
+            if (!anchorFound) {
+                require(
+                    anchorBlockId + _conf.maxAnchorHeightOffset >= uint48(block.number),
+                    AnchorIdTooSmall()
+                );
+            }
+
+            require(anchorBlockId > lastAnchorBlockId_, AnchorIdSmallerThanParent());
+
+            anchorBlockHashes_[k] = _rw.getBlobHash(anchorBlockId);
+            require(anchorBlockHashes_[k] != 0, ZeroAnchorBlockHash());
+
+            anchorFound = true;
+            lastAnchorBlockId_ = anchorBlockId;
+            k++;
+        }
+
+        // Ensure that if msg.sender is not the inboxWrapper, at least one block must
+        // have a non-zero anchor block id. Otherwise, delegate this validation to the
+        // inboxWrapper contract.
```

Comment:
Where in the "inboxWrapper" is this check done? Couldn't find in `TaikoWrapper` or `PreconfRouter` but maybe there are more "wrappers". Also, wonder if such check should even be done in the wrapper.

Reply:
fixed comment.